sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala

```diff
@@ -282,7 +282,27 @@ class MicroBatchExecution(
         /* Initialize committed offsets to a committed batch, which at this
          * is the second latest batch id in the offset log. */
         if (latestBatchId != 0) {
+          /* SPARK-38033: In some unexpected cases, commit and offset are inconsistent,
+            * and offset is not written into HDFS continuously, as follows:
+            *
+            * commits
+            * /tmp/streaming_xxxxxxxx/commits/113256
+            * /tmp/streaming_xxxxxxxx/commits/113257
+            * offsets
+            * /tmp/streaming_xxxxxxxx/offsets/113257
+            * /tmp/streaming_xxxxxxxx/offsets/113259
+            *
+            * When we start the streaming program, batch ${latestBatchId - 1} is 113258,
+            * but offsets 113258 doesn't exist, an exception will be thrown,resulting in
+            * the program cannot be started. As an improvement, we could probably do some
+            * simply analysis and give better error message. */
           val secondLatestOffsets = offsetLog.get(latestBatchId - 1).getOrElse {
+            logError(s"Please check the checkpoint, batch ${latestBatchId - 1} doesn't exist. " +
+              s"If the latest offset is $latestBatchId, the latest commit is ${latestBatchId - 2}" +
+              s" and offset ${latestBatchId - 1} doesn't exist. You can try to remove the offset" +
+              s" $latestBatchId and start over. If your query aims end-to-end exactly once" +
+              s" semantic, and you can also remove the output from the batch ${latestBatchId - 1}" +
```

Comment:
it is not that something "good to have" to fulfill the fault tolerance semantic. it is a requirement.

```suggestion
              s" semantic, you need to also remove the output from the batch ${latestBatchId - 1}" +
```

Reply:
That said, we need to also remove "if possible" in below line. If they can't do, it becomes at-least-once semantic.

Reply:
Yes, I got it, this is more accurate. Thank you for your suggestion.